function-sidecar/pkg/dispatcher/grpc/grpc.go

```diff
@@ -52,59 +58,157 @@ func (this *grpcDispatcher) Closed() <-chan struct{} {
 	return this.closed
 }
 
-func (this *grpcDispatcher) handleIncoming() {
+// A CorrelationStrategy groups together messages that belong to the same "spatial" stream (the "temporal" aspect
+// of grouping is taken care of by a WindowStrategy). Several invocations for different spatial groups may happen concurrently.
+type CorrelationStrategy func(message.Message) interface{}
+
+// A WindowingStrategyFactory is a function for creating a new instance of a (most certainly stateful) WindowingStrategy,
+// given a new, first message for a stream.
+type WindowingStrategyFactory func(message.Message) WindowingStrategy
+
+//
+type WindowingStrategy interface {
+	// ShouldClose notifies this strategy that a new message has just been sent to the function, resulting in the optional
+	// error passed as 2nd argument. This is a synchronous way of deciding whether or not a window should be closed.
+	// ShouldClose should return true if this strategy decides that the current stream should end.
+	ShouldClose(in message.Message, err error) bool
```

Comment:
It's still not clear whether `ShouldClose` needs to be passed the message that was used to create the `WindowingStrategy`. I'd prefer the factory to take no parameters to avoid this ambiguity, if that makes sense. (Sorry if I'm repeating myself, but I couldn't find the previous thread on this point or remember the conclusion.)

Reply:
ShouldClose doesn't take the message that was passed to the strategy (ie the *first* one), but rather the last (ie most recent) one. They're only the same in the case of window-of-size-one.

The factory takes the first message to ease the case of remembering eg when a stream was created

Reply:
Ok. May I suggest a small improvement to the comment? The phrase "a new message" might be taken to imply "a message different from that which was passed to the factory". I suggest something along these lines (which also clarified the _requirement_ to call this method - for example, it's not acceptable simply to call this method from time to time):

> This method should be called with each message that is sent to the function, including the 
> message that was passed to the `WindowingStrategyFactory` to construct the 
> `WindowingStrategy`.